### PR TITLE
feature 299 update login with tonomy

### PR DESCRIPTION
**Pull request by sadiabbasi**

### CODE REVIEW

File: src/api/externalUser.ts [hunk 1]
```diff
@@ -251,6 +262,7 @@ export class ExternalUser {
             const payload: LoginRequestsMessagePayload = {
                 requests: [loginRequest],
```
Comment (sadiabbasi) [inline src/api/externalUser.ts:263]: here i am just thinking do we need to add the loginRequest and datasharingRequest? i am confused with this part

Comment (theblockstalk) [reply src/api/externalUser.ts:263]: yes, we probably want `requests: [loginRequest, dataSharingRequest]`

File: src/sdk/util/request.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 import { VCWithTypeType, VerifiableCredentialOptions, VerifiableCredentialWithType } from './ssi/vc';
 import { Issuer } from '@tonomy/did-jwt-vc';
+import { DataSharingRequest } from './dataSharingRequest';
 
 export type LoginRequestPayload = {
     randomString: string;
     origin: string;
     publicKey: PublicKey;
     callbackPath: string;
+    dataSharingRequest?: DataSharingRequest;
```
Comment (theblockstalk) [inline src/sdk/util/request.ts:11]: This should not be here, instead we should have it as part of `LoginRequestsMessagePayload,`

https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/master/src/sdk/services/communication/message.ts#L118C1-L120C3
```ts
export type LoginRequestsMessagePayload = {
    requests: Request[];
};
```

**Review by theblockstalk — COMMENTED** [review]
Update the Integration test so that it shows this flow

**Comment by sadiabbasi** [comment]
> Update the Integration test so that it shows this flow

https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/issues/300 this task have to update integration test

File: src/sdk/controllers/userApps.ts [hunk 3]
```diff
@@ -177,38 +181,43 @@ export class UserApps {
...
                 });
-                requiresLogin = false;
-            } catch (e) {
-                if (e instanceof SdkError && e.code === SdkErrors.UserNotLoggedInWithThisApp) {
-                    // Never consented
-                    requiresLogin = true;
-                } else {
-                    throw e;
-                }
+            } else if (request instanceof DataSharingRequest) {
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:218]: i dont think we need this condition

File: src/sdk/controllers/userApps.ts [hunk 4]
```diff
@@ -177,38 +181,43 @@ export class UserApps {
      * @param {LoginRequest[]} requests - Array of login requests to check
      * @returns {Promise<CheckedRequest[]>} - Array of requests that have been verified and had authorization checked
      */
-    async checkRequests(requests: LoginRequest[]): Promise<CheckedRequest[]> {
+    async checkRequests(requests: Request[]): Promise<CheckedRequest[]> {
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:184]: maybe this should be renamed `checkLoginRequests()` ?

File: src/sdk/controllers/userApps.ts [hunk 5]
```diff
@@ -273,14 +287,16 @@ export class UserApps {
...
-        if (!base64UrlPayload) throwError("payload parameter doesn't exists", SdkErrors.MissingParams);
+        if (!base64UrlPayload) throwError("payload parameter doesn't exist", SdkErrors.MissingParams);
 
         const parsedPayload = base64UrlToObj(base64UrlPayload);
 
         if (!parsedPayload || !parsedPayload.requests)
             throwError('No requests found in payload', SdkErrors.MissingParams);
 
-        const loginRequests = parsedPayload.requests.map((r: string) => new LoginRequest(r));
+        const loginRequests = parsedPayload.requests.map((r: string) => {
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:297]: this may need to filter based on the request type

File: src/sdk/controllers/userApps.ts [hunk 6]
```diff
@@ -337,12 +353,29 @@ export class UserApps {
 
         const referrer = new URL(docReferrer);
 
-        const myRequest = verifiedRequests.find((r) => r.getPayload().origin === referrer.origin);
+        const myRequest = verifiedRequests.find((r) => {
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:356]: const myRequest = verifiedRequests.find((r) => r instanceof LoginRequest && r.getPayload().origin === referrer.origin;
        });

File: src/sdk/controllers/userApps.ts [hunk 7]
```diff
@@ -337,12 +353,29 @@ export class UserApps {
...
+            }
+
+            return;
+        });
 
         if (!myRequest) {
             const msg =
-                `No origins from: ${verifiedRequests.map((r) => r.getPayload().origin)} ` +
-                `match referrer: ${referrer.origin}`;
+                `No origins from: ${verifiedRequests.map((r) => {
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:369]: Do a verifiedRequest.find(find login requests).map(return origin)

**Review by theblockstalk — COMMENTED** [review]
I think you also ned to modify the `LoginRequestResponseMessage` type, and so that it turns the username when an appropriate DataSharingRequest is found that requests it

File: src/api/externalUser.ts [hunk 8]
```diff
@@ -303,13 +319,27 @@ export class ExternalUser {
...
+                return false;
+            });
             const keyFromStorage = await keyManager.getKey({ level: KeyManagerLevel.BROWSER_LOCAL_STORAGE });
 
             if (!loginRequest) throwError('No login request found for this origin', SdkErrors.OriginMismatch);
 
-            if (loginRequest.publicKey.toString() !== keyFromStorage.toString()) {
-                throwError('Key in request does not match', SdkErrors.KeyNotFound);
+            if (loginRequest.getType() === 'LoginRequest') {
+                const payload = loginRequest.getPayload() as LoginRequestPayload;
```
Comment (theblockstalk) [inline src/api/externalUser.ts:338]: instead of using `as`, you could do
`const payload = new LoginRequest(loginRequest).getPayload()`

File: src/sdk/controllers/userApps.ts [hunk 9]
```diff
@@ -177,38 +181,41 @@ export class UserApps {
...
 
         await UserApps.verifyRequests(requests);
 
         for (const request of requests) {
-            const payload = request.getPayload();
-            const app = await App.getApp(payload.origin);
-
-            let requiresLogin = true;
+            if (request.getType() === 'LoginRequest') {
+                const payload = request.getPayload() as LoginRequestPayload;
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:191]: same comment as above

File: src/api/externalUser.ts (hunk 8 above)
Comment (theblockstalk) [reply src/api/externalUser.ts:338]: this way we KNOW we are future proofed for type safety

File: src/sdk/controllers/userApps.ts [hunk 10]
```diff
@@ -273,14 +290,30 @@ export class UserApps {
...
-        if (!base64UrlPayload) throwError("payload parameter doesn't exists", SdkErrors.MissingParams);
+        if (!base64UrlPayload) throwError("payload parameter doesn't exist", SdkErrors.MissingParams);
 
         const parsedPayload = base64UrlToObj(base64UrlPayload);
 
         if (!parsedPayload || !parsedPayload.requests)
             throwError('No requests found in payload', SdkErrors.MissingParams);
 
-        const loginRequests = parsedPayload.requests.map((r: string) => new LoginRequest(r));
+        // const loginRequests = parsedPayload.requests.map((r: string) => new LoginRequest(r));
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:300]: remove comment

File: src/sdk/controllers/userApps.ts [hunk 11]
```diff
@@ -273,14 +290,30 @@ export class UserApps {
...
+        parsedPayload.requests = parsedPayload.requests.filter((request: string) => request !== null);
```
Comment (theblockstalk) [inline src/sdk/controllers/userApps.ts:301]: should this be done above L297?

Comment (theblockstalk) [reply src/sdk/controllers/userApps.ts:301]: that would then catch an edge case where `parsedPayload.requests` contained an array of null objects

Comment (theblockstalk) [reply src/sdk/controllers/userApps.ts:301]: that would then catch an edge case where `parsedPayload.requests` contained an array of null objects

File: test/controllers/app.test.ts [hunk 12]
```diff
@@ -49,12 +49,13 @@ describe('logging in', () => {
         });
 
         const result = await UserApps.onRedirectLogin();
+        const loginRequests = result.getPayload() as LoginRequestPayload;
```
Comment (theblockstalk) [inline test/controllers/app.test.ts:52]: see comment above

**Review by theblockstalk — APPROVED** [review]
